dalle2_pytorch/train_diffusion_prior.py

```diff
@@ -0,0 +1,160 @@
+from embedding_reader import EmbeddingReader
+from dalle2_pytorch import DiffusionPriorNetwork
+import numpy as np
+import math
+from tqdm import tqdm
+import torch
+from torch import nn
+import wandb
+os.environ["WANDB_SILENT"] = "true"
+
+def train(image_embed_dim,
+          image_embed_url,
+          text_embed_url,
+          batch_size,
+          train_percent,
+          val_percent,test_percent,
+          num_epochs,
+          loss_type,
+          condition_on_text_encodings,
+          device,
+          learning_rate=0.01):
+    # DiffusionPriorNetwork 
+    prior_network = DiffusionPriorNetwork( dim = image_embed_dim, depth = 6, dim_head = 64, heads = 8).to(device)
+    
+    # DiffusionPrior with text embeddings and image embeddings pre-computed
+    diffusion_prior = DiffusionPrior( net = prior_network, clip = None, image_embed_dim = image_embed_dim, 
+                                     timesteps = 100, cond_drop_prob = 0.2, loss_type = loss_type,
+                                     condition_on_text_encodings = condition_on_text_encodings).to(device)
+    # Gtext_reader image and text embeddings from the servers
+    print("==============Downloading embeddings - image and text====================")
+    image_reader = EmbeddingReader(embeddings_folder=image_embed_url, file_format="npy")
+    text_reader = EmbeddingReader(embeddings_folder=text_embed_url, file_format="npy")
+
+    ### Training code ###
+    optimizer = torch.optim.SGD(diffusion_prior.parameters(), lr = learning_rate)
+    epochs = num_epochs
+    min_valid_loss = np.inf
+    for e in range(epochs):
+        train_loss = 0.0 
+        print("Training loop - epoch number ",e)
+        train_set_size = int(train_percent*image_reader.count)
+        for emb_images,emb_text in zip(image_reader(batch_size=batch_size, start=0, end=train_set_size),
+                text_reader(batch_size=batch_size, start=0, end=train_set_size)):
+            emb_images = list(emb_images)
+            emb_text = list(emb_text)
+            print(emb_images[0].shape,emb_text[0].shape)
+            if torch.cuda.is_available():
+                emb_images[0] = torch.tensor(emb_images[0]).to(device)
+                emb_text[0] = torch.tensor(emb_text[0]).to(device)
+            optimizer.zero_grad()
+            loss = diffusion_prior(text_embed = emb_text[0],image_embed = emb_images[0])
+            loss.backward()
+            # Log to wandb
+            wandb.log({"Training (mse)": loss})
+            optimizer.step()
+            print("Training mse = ",loss.item())
+            train_loss+=loss.item()
+
+        print("Validation loop - epoch number ",e)
+        with torch.no_grad():
+            valid_loss = 0.0
+            val_set_size = int(val_percent*image_reader.count)
+            start = train_set_size
+            end=start+val_set_size
+            for emb_images,emb_text in zip(image_reader(batch_size=batch_size, start=start, end=end),
```

Comment:
is this a separate set that isn't seen during training? (for your validation loop)

Reply:
I understand this is controlled by the start and end parameters here
The way embedding reader work is you give it a bunch of npy files (that can be located locally, at s3, at http or whatever else filesystem fsspec support), a start an end and a batch size and then it gives you batches

the start and end can indeed be used to split the dataset

Reply:
ohh got it, perfect :100: !